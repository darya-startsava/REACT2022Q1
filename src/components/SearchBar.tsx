import React, { ChangeEvent } from 'react';

interface InputProps extends React.HTMLProps<HTMLInputElement> {}

export default class SearchBar extends React.Component<InputProps, { value: string }> {
  constructor(props: InputProps) {
    super(props);
    this.state = {
      value: localStorage.getItem('value') || '',
    };
    this.handleChange = this.handleChange.bind(this);
  }

  componentWillUnmount() {
    localStorage.setItem('value', this.state.value);
  }

  handleChange(event: ChangeEvent<HTMLInputElement>) {
    this.setState({ value: event.target.value });
  }

  render() {
    return (
      <label className="my-2">
        <input
          type="text"
          placeholder="Search bar"
          value={this.state.value}
          onChange={this.handleChange}
        />
      </label>
    );
  }
}
