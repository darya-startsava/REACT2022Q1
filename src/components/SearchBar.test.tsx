import { render, screen } from '@testing-library/react';
import SearchBar from './SearchBar';
import '@testing-library/jest-dom/extend-expect';
import userEvent from '@testing-library/user-event';

describe('Search bar', () => {
  beforeEach(() => {
    class LocalStorageMock {
      store: Record<string, string>;
      length = 0;
      constructor() {
        this.store = {};
      }

      key(index: number) {
        return Object.keys(this.store)[index];
      }

      clear() {
        this.store = {};
      }

      getItem(key: string) {
        return this.store[key] || null;
      }

      setItem(key: string, value: string) {
        this.store[key] = String(value);
      }

      removeItem(key: string) {
        delete this.store[key];
      }
    }

    Object.defineProperty(global, '_localStorage', {
      value: new LocalStorageMock(),
      writable: true,
    });
  });

  it('search bar value update on change', () => {
    render(<SearchBar />);
    const input = screen.queryByPlaceholderText('Search bar') as HTMLInputElement;
    userEvent.type(input, 'test1');
    expect(input.value).toBe('test1');
  });

  it('search bar value show value from local storage', () => {
    global.localStorage.setItem('value', 'test2');
    const { container } = render(<SearchBar />);
    expect(container).toMatchSnapshot();
  });

  it('search bar value is saved to local storage when unmount', () => {
    const { unmount } = render(<SearchBar />);
    const input = screen.queryByPlaceholderText('Search bar') as HTMLInputElement;
    userEvent.type(input, 'test3');
    unmount();
    expect(global.localStorage.getItem('value')).toBe('test3');
  });
});
