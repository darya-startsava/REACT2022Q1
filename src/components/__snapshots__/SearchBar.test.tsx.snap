// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Search bar should show the value from local storage 1`] = `
<div>
  <label
    class="my-2"
  >
    <input
      placeholder="Search bar"
      type="text"
      value="test1"
    />
  </label>
</div>
`;
